File: impl/dart/lib/src/mron.dart
import "dart:io";

import "_mbf_lite.dart";
import "error.dart";

class Mron {
  static Object? parseString(String source, {bool allowExec = false, Set<String> profiles = const {}}) {
    if (allowExec) {
      throw MakrellFormatException("MRON executable embeds are not implemented in the Dart track.");
    }

    final stream = MbfLiteTokenStream(tokenizeMbfLite(source));
    final values = <Object?>[];
    while (!stream.isAtEnd) {
      values.add(_parseValue(stream, profiles: profiles));
    }

    if (values.isEmpty) return null;
    if (values.length == 1) return values.first;
    if (values.length.isOdd) {
      throw MakrellFormatException("Illegal number (${values.length}) of root level expressions");
    }
    return _pairsToMap(values);
  }

  static Object? parseFile(String path, {bool allowExec = false, Set<String> profiles = const {}}) {
    return parseString(File(path).readAsStringSync(), allowExec: allowExec, profiles: profiles);
  }

  static String writeString(Object? value) {
    return _writeValue(value);
  }

  static Object? _parseValue(MbfLiteTokenStream stream, {required Set<String> profiles}) {
    final token = stream.consume();
    switch (token.kind) {
      case MbfLiteTokenKind.lBracket:
        final items = <Object?>[];
        while (!stream.match(MbfLiteTokenKind.rBracket)) {
          if (stream.isAtEnd) {
            throw MakrellFormatException("Unterminated MRON list.");
          }
          items.add(_parseValue(stream, profiles: profiles));
        }
        return items;
      case MbfLiteTokenKind.lBrace:
        final values = <Object?>[];
        while (!stream.match(MbfLiteTokenKind.rBrace)) {
          if (stream.isAtEnd) {
            throw MakrellFormatException("Unterminated MRON object.");
          }
          values.add(_parseValue(stream, profiles: profiles));
        }
        return _pairsToMap(values);
      case MbfLiteTokenKind.identifier:
      case MbfLiteTokenKind.number:
      case MbfLiteTokenKind.string:
        return parseCoreScalar(token, profiles: profiles);
      default:
        throw MakrellFormatException("Unexpected token '${token.lexeme}' in MRON.");
    }
  }

  static Map<String, Object?> _pairsToMap(List<Object?> values) {
    if (values.isNotEmpty && values.length.isOdd) {
      throw MakrellFormatException("Odd pair count in object-like MRON form.");
    }

    final result = <String, Object?>{};
    for (var index = 0; index < values.length; index += 2) {
      final key = values[index];
      if (key is! String) {
        throw MakrellFormatException("MRON object keys must be string-compatible scalars.");
      }
      result[key] = values[index + 1];
    }
    return result;
  }

  static String _writeValue(Object? value) {
    if (value is Map) {
      if (value.isEmpty) return "{}";
      final parts = <String>[];
      value.forEach((key, nested) {
        parts.add("${scalarToSource(key.toString())} ${_writeValue(nested)}");
      });
      return "{ ${parts.join(" ")} }";
    }
    if (value is Iterable && value is! String) {
      return "[${value.map(_writeValue).join(" ")}]";
    }
    return scalarToSource(value);
  }
}

File: impl/jvm/src/main/java/dev/makrell/formats/Mron.java
package dev.makrell.formats;

import dev.makrell.formats.internal.MiniMbf;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class Mron {
    private Mron() {
    }

    public static Object parseString(String source) {
        return parseString(source, Set.of(), false);
    }

    public static Object parseFile(String path) {
        try {
            return parseString(Files.readString(Paths.get(path), StandardCharsets.UTF_8));
        } catch (IOException ex) {
            throw new MakrellFormatException("Could not read MRON file: " + path);
        }
    }

    public static String writeString(Object value) {
        return writeValue(value);
    }

    public static Object parseString(String source, Set<String> profiles, boolean allowExec) {
        if (allowExec && source.contains("{$")) {
            throw new MakrellFormatException("Executable embeds are not supported in the JVM track yet.");
        }
        List<MiniMbf.Node> roots = MiniMbf.parse(source);
        if (roots.isEmpty()) {
            return null;
        }
        if (roots.size() == 1) {
            return convertNode(roots.get(0), profiles);
        }
        if (roots.size() % 2 != 0) {
            throw new MakrellFormatException("Illegal number (" + roots.size() + ") of root level expressions for MRON object.");
        }
        return convertPairs(roots, profiles);
    }

    private static Object convertNode(MiniMbf.Node node, Set<String> profiles) {
        if ("scalar".equals(node.kind)) {
            return convertScalar(node.text, node.quoted, profiles);
        }
        if ("square".equals(node.kind)) {
            List<Object> values = new ArrayList<>();
            for (MiniMbf.Node child : node.children) {
                values.add(convertNode(child, profiles));
            }
            return values;
        }
        if ("brace".equals(node.kind)) {
            return convertPairs(node.children, profiles);
        }
        throw new MakrellFormatException("Unsupported MRON node kind: " + node.kind);
    }

    private static Map<String, Object> convertPairs(List<MiniMbf.Node> nodes, Set<String> profiles) {
        if (nodes.size() % 2 != 0) {
            throw new MakrellFormatException("Odd pair count in MRON object.");
        }
        Map<String, Object> result = new LinkedHashMap<>();
        for (int i = 0; i < nodes.size(); i += 2) {
            Object keyValue = convertNode(nodes.get(i), profiles);
            result.put(String.valueOf(keyValue), convertNode(nodes.get(i + 1), profiles));
        }
        return result;
    }

    private static Object convertScalar(String text, boolean quoted, Set<String> profiles) {
        if (quoted) {
            if (profiles.contains("extended-scalars") && text.endsWith("dt")) {
                return tryParseDateTime(text.substring(0, text.length() - 2));
            }
            return text;
        }
        if ("null".equals(text)) {
            return null;
        }
        if ("true".equals(text)) {
            return Boolean.TRUE;
        }
        if ("false".equals(text)) {
            return Boolean.FALSE;
        }
        if (text.matches("-?\\d+")) {
            try {
                return Integer.valueOf(text);
            } catch (NumberFormatException ex) {
                return Long.valueOf(text);
            }
        }
        if (text.matches("-?\\d+\\.\\d+")) {
            return Double.valueOf(text);
        }
        if (profiles.contains("extended-scalars") && text.matches("-?\\d+(\\.\\d+)?k")) {
            return Double.valueOf(text.substring(0, text.length() - 1)) * 1000.0;
        }
        return text;
    }

    private static Object tryParseDateTime(String text) {
        try {
            return OffsetDateTime.parse(text);
        } catch (RuntimeException ignored) {
        }
        try {
            return LocalDate.parse(text);
        } catch (RuntimeException ignored) {
        }
        return text;
    }

    private static String writeValue(Object value) {
        if (value == null) {
            return "null";
        }
        if (value instanceof Boolean || value instanceof Number) {
            return String.valueOf(value);
        }
        if (value instanceof String) {
            String text = (String) value;
            return isIdentifierLike(text) ? text : quote(text);
        }
        if (value instanceof Map<?, ?>) {
            List<String> parts = new ArrayList<>();
            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
                parts.add(writeValue(String.valueOf(entry.getKey())));
                parts.add(writeValue(entry.getValue()));
            }
            return "{ " + String.join(" ", parts) + " }";
        }
        if (value instanceof List<?>) {
            List<String> parts = new ArrayList<>();
            for (Object item : (List<?>) value) {
                parts.add(writeValue(item));
            }
            return "[" + String.join(" ", parts) + "]";
        }
        throw new MakrellFormatException("Unsupported MRON value for serialisation: " + value.getClass().getName());
    }

    private static boolean isIdentifierLike(String value) {
        return value.matches("[A-Za-z_][A-Za-z0-9_\\-]*");
    }

    private static String quote(String value) {
        return "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }
}
